File: rpn/python/rpn.py
#!/usr/bin/python

from __future__ import print_function

from sys import stdin

ops = {}

def signalError(string):
    print("*** ERROR: " + string)

class Operation (object):
    def __init__ (self, arity, doc, op):
        self.arity = arity
        self.doc = doc
        self.op = op

    def __call__ (self, stack):
        args = []
        if len(stack) < self.arity:
            signalError("stack underflow")
        else:
            for i in range(self.arity):
                args.append(stack.pop())
            args.reverse()
            self.op(stack, *args)

# given an arity, return a decorator for that arity
def operation(name, arity, doc):
    def decorate(op):
       ops[name] = Operation(arity, doc, op)
       return ops[name]
    return decorate

@operation ('.', 1, 'display the top value on the stack')
def op_show (stack, x):
    print(x)
    stack.append(x)

@operation('#', 0, 'display number of values on the stack')
def op_size (stack):
    print(len(stack))

@operation('+', 2, 'replace the top two values on the stack with their sum')
def op_plus (stack, x, y):
    stack.append(x + y)

@operation('-', 2, 'replace the top two values on the stack with their difference')
def op_minus (stack, x, y):
    stack.append(x - y)

@operation('*', 2, 'replace the top two values on the stack with their product')
def op_times (stack, x, y):
    stack.append(x * y)

@operation('/', 2, 'replace the top two values on the stack with their quotient')
def op_divby (stack, x, y):
    stack.append(x / y)

@operation('^', 2, 'replace the top two values on the stack, x and y, with x to the yth power')
def op_expt (stack, x, y):
    stack.append(x ** y)

@operation('drop', 1, 'remove the top value from the stack')
def op_drop (stack, x, y):
    pass

@operation('dup', 1, 'duplicate the top value on the stack')
def op_dup (stack, x):
    stack.append(x)
    stack.append(x)

@operation('swap', 2, 'swap the top two values on the stack')
def op_swap (stack, x, y):
    stack.append(y)
    stack.append(x)

@operation ('help', 0, 'display this help')
def op_help (stack):
    print(str(len(ops.keys())) + " Commands:")
    for c in ops.keys():
        print(c + " -- " + ops[c].doc)

stack = []

def action(string):
    if string in ops:
        ops[string](stack)
    else:
        try:
            stack.append(float(string))
        except ValueError:
            signalError("unknown operation")

while True:
    print("> ", end='')
    string = stdin.readline()
    if string == '':
        break
    elif string == '\n':
        continue
    else:
        action(string[:-1])

print('')

File: rpn/ocaml/rpn.ml
(* we use lists instead of the Stack stdlib, as we need a functional structure *)

type op = {
    doc : string ;
    action : float list -> float list
  } ;;

let dict = Hashtbl.create 32 ;;

let keys d = Hashtbl.fold (fun k v acc -> k :: acc) d [] ;;

let make_op name d a =
  Hashtbl.add dict name { doc=d; action=a } ;;

let signal_error s =
  print_endline s ;;

let underflow stack =
  signal_error "stack underflow"; stack ;;

let show_help () =
  Printf.printf "%d Commands:\n" ( Hashtbl.length dict ) ;
  let keys = List.sort compare (keys dict) in
  let print_cmd k = Printf.printf "%s -- %s\n" k (Hashtbl.find dict k).doc in
  List.iter print_cmd keys ;;

make_op "." "display top value on the stack" (
function ( x :: stack ) -> Printf.printf "%f\n" x ; x :: stack 
  | stack -> underflow stack )
;;
make_op "#" "display number of values on the stack" ( 
function stack -> print_int ( List.length stack ); print_newline () ; stack )
;;

make_op "+" "replace top two values on the stack with their sum" (
function ( y :: x :: stack ) -> ( x +. y ) :: stack
  | stack -> underflow stack )
;;
make_op "-" "replace top two values on the stack with their difference" (
function ( y :: x :: stack ) -> ( x -. y ) :: stack
  | stack -> underflow stack ) 
;;
make_op "*" "replace top two values on the stack with their product" (
function ( y :: x :: stack ) -> ( x *. y ) :: stack
  | stack -> underflow stack )
;;
make_op "/" "replace top two values on the stack with their quotient" (
function ( y :: x :: stack ) -> ( x /. y ) :: stack
  | stack -> underflow stack )
;;
make_op "^" "replace top two values on the stack, x and y, with x to the yth power" (
function ( y :: x :: stack ) -> ( x ** y ) :: stack
  | stack -> underflow stack )
;;

make_op "drop" "remove top value from the stack" (
function ( x :: stack ) -> stack
  | stack -> underflow stack )
;;
make_op "dup" "duplicate top value on the stack" (
function ( x :: stack ) -> x :: x :: stack
  | stack -> underflow stack )
;;
make_op "swap" "swap top two values on the stack" (
function ( y :: x :: stack ) -> x :: y :: stack
  | stack -> underflow stack )
;;

make_op "help" "display this help" ( function stack -> show_help () ; stack ) ;;


let act stack s =
  try
    (Hashtbl.find dict s).action stack ;
  with
    Not_found ->
      try
        ( float_of_string s ) :: stack ;
      with
        Failure _ -> signal_error ("unknown operation: " ^ s) ; stack
;;
  
let rec rpn stack =
  try
    print_string "> " ;
    flush stdout ;
    let s = read_line() in
    let newstack = act stack s in
    rpn newstack
  with
    End_of_file ->
      print_endline "" ;; (* XXX not sure why print_newline doesn't do it *)
  
rpn [ ];;
